File: src/Core/Seguranca/Ataques/AtaqueAvalanche.php
<?php

namespace Application\Core\Seguranca\Ataques;

use Application\Core\Seguranca\AlvoCriptografico;
use Application\Core\Seguranca\AtaqueInterface;
use Application\Core\Seguranca\ResultadoAtaque;
use Application\Core\Seguranca\SkipAtaqueException;

/**
 * Mudar 1 bit do IV deveria, em média, mudar ~50% dos bits do keystream
 * gerado (efeito avalanche). Números muito abaixo disso indicam difusão
 * fraca - foi o que achamos quando o número de rodadas de mistura era
 * baixo demais numa das versões anteriores.
 */
class AtaqueAvalanche implements AtaqueInterface
{
    public function __construct(
        private int $amostras = 3000,
        private int $tamanhoBloco = 32,
        private float $limiteMedia = 0.45,
        private float $limitePiorCaso = 0.30,
    ) {
    }

    public function nome(): string
    {
        return 'Efeito avalanche';
    }

    public function executar(AlvoCriptografico $alvo): ResultadoAtaque
    {
        $chave = $alvo->chaveDeTeste();
        $tamanhoIv = $alvo->tamanhoIv();

        $primeiro = $alvo->gerarKeystreamBruto($chave, random_bytes($tamanhoIv), 'enc', $this->tamanhoBloco);
        if ($primeiro === null) {
            throw new SkipAtaqueException('Alvo não expõe gerarKeystreamBruto.');
        }

        $valores = [];
        for ($t = 0; $t < $this->amostras; $t++) {
            $iv1 = random_bytes($tamanhoIv);
            $iv2 = $iv1;
            $pos = random_int(0, $tamanhoIv - 1);
            $iv2[$pos] = chr(ord($iv2[$pos]) ^ (1 << random_int(0, 7)));

            $ks1 = $alvo->gerarKeystreamBruto($chave, $iv1, 'enc', $this->tamanhoBloco);
            $ks2 = $alvo->gerarKeystreamBruto($chave, $iv2, 'enc', $this->tamanhoBloco);

            $valores[] = $this->bitsDiferentes($ks1, $ks2) / ($this->tamanhoBloco * 8);
        }

        sort($valores);
        $n = count($valores);
        $media = array_sum($valores) / $n;
        $piorCaso = $valores[0];

        $vulneravel = $media < $this->limiteMedia || $piorCaso < $this->limitePiorCaso;

        return new ResultadoAtaque(
            $this->nome(),
            vulneravel: $vulneravel,
            severidade: $vulneravel ? 'alta' : 'info',
            detalhes: sprintf(
                'média=%.1f%%, pior caso=%.1f%%, percentil 1%%=%.1f%% (limites: média>=%.0f%%, pior>=%.0f%%)',
                $media * 100,
                $piorCaso * 100,
                $valores[(int) ($n * 0.01)] * 100,
                $this->limiteMedia * 100,
                $this->limitePiorCaso * 100,
            ),
            dados: ['media' => $media, 'pior_caso' => $piorCaso],
        );
    }

    private function bitsDiferentes(string $a, string $b): int
    {
        $diff = 0;
        for ($i = 0; $i < strlen($a); $i++) {
            $diff += substr_count(decbin(ord($a[$i]) ^ ord($b[$i])), '1');
        }
        return $diff;
    }
}

File: packages/go/internal/seguranca/ataques/ataque_avalanche.go
package ataques

import (
	"fmt"
	"sort"

	"criptografia/internal/seguranca"
)

// OpcoesAvalanche configura o AtaqueAvalanche.
type OpcoesAvalanche struct {
	Amostras       int
	TamanhoBloco   int
	LimiteMedia    float64
	LimitePiorCaso float64
}

// AtaqueAvalanche mede a difusão de 1 bit do IV no keystream gerado.
type AtaqueAvalanche struct {
	amostras       int
	tamanhoBloco   int
	limiteMedia    float64
	limitePiorCaso float64
}

// NovoAtaqueAvalanche cria o ataque. Sem argumento, usa amostras = 3000,
// tamanhoBloco = 32, limiteMedia = 0.45 e limitePiorCaso = 0.3.
func NovoAtaqueAvalanche(opcoes ...OpcoesAvalanche) *AtaqueAvalanche {
	o := OpcoesAvalanche{Amostras: 3000, TamanhoBloco: 32, LimiteMedia: 0.45, LimitePiorCaso: 0.3}
	if len(opcoes) > 0 {
		o = opcoes[0]
	}
	return &AtaqueAvalanche{
		amostras:       o.Amostras,
		tamanhoBloco:   o.TamanhoBloco,
		limiteMedia:    o.LimiteMedia,
		limitePiorCaso: o.LimitePiorCaso,
	}
}

// Nome devolve o nome exibido no relatório.
func (a *AtaqueAvalanche) Nome() string {
	return "Efeito avalanche"
}

// Executar roda o ataque.
func (a *AtaqueAvalanche) Executar(alvo seguranca.AlvoCriptografico) (seguranca.ResultadoAtaque, error) {
	chave := []byte(alvo.ChaveDeTeste())
	tamanhoIv := alvo.TamanhoIv()

	valores := make([]float64, 0, a.amostras)
	for t := 0; t < a.amostras; t++ {
		iv1 := seguranca.RandomBytes(tamanhoIv)
		iv2 := append([]byte(nil), iv1...)
		pos := seguranca.RandomInt(0, tamanhoIv-1)
		iv2[pos] ^= byte(1 << seguranca.RandomInt(0, 7))

		ks1 := alvo.GerarKeystreamBruto(chave, iv1, "enc", a.tamanhoBloco)
		ks2 := alvo.GerarKeystreamBruto(chave, iv2, "enc", a.tamanhoBloco)

		valores = append(valores, float64(seguranca.BitsDiferentes(ks1, ks2))/float64(a.tamanhoBloco*8))
	}

	sort.Float64s(valores)
	n := len(valores)
	media := 0.0
	for _, v := range valores {
		media += v
	}
	media /= float64(n)
	piorCaso := valores[0]

	vulneravel := media < a.limiteMedia || piorCaso < a.limitePiorCaso
	severidade := seguranca.SevInfo
	if vulneravel {
		severidade = seguranca.SevAlta
	}

	percentil := valores[int(float64(n)*0.01)]

	return seguranca.ResultadoAtaque{
		NomeAtaque: a.Nome(),
		Vulneravel: vulneravel,
		Severidade: severidade,
		Detalhes: fmt.Sprintf(
			"média=%.1f%%, pior caso=%.1f%%, percentil 1%%=%.1f%% (limites: média>=%.0f%%, pior>=%.0f%%)",
			media*100, piorCaso*100, percentil*100, a.limiteMedia*100, a.limitePiorCaso*100,
		),
		Dados: map[string]any{"media": media, "pior_caso": piorCaso},
	}, nil
}
